hyperspy/io_plugins/emd.py

```diff
@@ -404,14 +421,581 @@ def log_info(self):
+            stream = FeiSpectrumStream(stream_data, **kwargs)
+
+            acquisition_key = '{}/AcquisitionSettings'.format(key)
+            acquisition_settings_group = self.spectrum_stream_group[acquisition_key]
+            stream.parse_acquisition_settings(acquisition_settings_group)
+
+            metadata_group = self.spectrum_stream_group[key]
+            stream.parse_metadata(metadata_group)
+
+            stream.get_spectrum_image()
+            stream_list.append(stream)
+
+        return stream_list
+
+    def read_streams(self, individual_detector=False):
+        self.stream_list = self._read_streams()
+
+        self.frame_number = self.stream_list[0].frame_number
+        if individual_detector is False:
+            self.sum_spectrum_image = self.stream_list[0].spectrum_image
+            for stream in self.stream_list[1:]:
+                self.sum_spectrum_image += stream.spectrum_image
+
+    def get_SI_shape(self):
+        return self.stream_list[0].spectrum_image.shape
+
+    def get_pixelsize_offset_unit(self, stream_index=0):
+        om_br = self.stream_list[stream_index].original_metadata['BinaryResult']
+        return om_br['PixelSize'], om_br['Offset'], om_br['PixelUnitX']
+
+
+class FeiSpectrumStream(object):
+
+    """
+    Below some information we have got from FEI:
+    'The SI data is stored as a spectrum stream, ‘65535’ means next pixel,
+    other numbers mean a spectrum count in that bin for that pixel.
+    For the size of the spectrum image and dispersion you have to look in
+    AcquisitionSettings.
+    The spectrum image cube itself stored in a compressed format, that is
+    not easy to decode.'
+    """
+
+    def __init__(self, stream_data, shape, energy_rebin=1, first_frame=0,
+                 last_frame=None, individual_frame=False, data_dtype=None):
+        self.stream_data = stream_data
+        self.shape = shape
+        self.energy_rebin = energy_rebin
+        self.first_frame = first_frame
+        self.last_frame = last_frame
+        self.individual_frame = individual_frame
+        self.data_dtype = data_dtype
+
+    def parse_metadata(self, stream_group):
+        self.original_metadata = _parse_sub_data_group_metadata(stream_group)
+
+    def _add_imported_parameters_to_original_metadata(self):
+        self.original_metadata['ImportedDataParameter'] = {
+            'First_frame': self.first_frame,
+            'Last_frame': self.last_frame,
+            'Frame_number': self.frame_number}
+
+    def parse_acquisition_settings(self, acquisition_settings_group):
+        acquisition_settings = json.loads(
+            acquisition_settings_group.value[0].decode('utf-8'))
+        self.bin_count = int(acquisition_settings['bincount'])
+        if self.data_dtype is None:
+            self.data_dtype = acquisition_settings['StreamEncoding']
+        return acquisition_settings
+
+    def get_spectrum_image(self):
+        stream = self.stream_data
```

Comment:
Starting from Velox 2.1, in the same group as the stream, we also save a "FrameLocationTable" dataset. This contains the offset of the beginning of each frame within the spectrum stream. This can used to quickly load subset range of frames from the stream.

Reply:
Very good that you tell us this, this will be very useful to load the data lazily!!